feat: unify ModuleSystem and DescriptorSystem [BD-13]

kaustavb12 opened the pull request:
<!--

🎉🎉 Olive has been released! 🎉🎉

🫒🫒
🫒🫒🫒🫒         🫒 Note: Olive is in support. Fixes you make on master may still
    🫒🫒🫒🫒     be needed on Olive. If so, make another pull request against the
🫒🫒🫒🫒         open-release/olive.master branch, or ping @nedbat for help or questions.
🫒🫒

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply. You may link to information rather than copy it.
More details about the template are at https://github.com/openedx/open-edx-proposals/pull/180
(link will be updated when that document merges)
-->

## Description

Work in progress PR to unify ModuleSystem and DescriptorSystem

## Supporting information

OpenCraft internal ticket [BB-5967](https://tasks.opencraft.com/browse/BB-5967)

## Testing instructions

Please provide detailed step-by-step instructions for testing this change.

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

Include anything else that will help reviewers and consumers understand the change.
- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, security, or accessibility.
- If your [database migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) can't be rolled back easily.

kaustavb12 committed:
chore: move ModuleSystemShim above DescriptorSystem

openedx-webhooks commented:
Thanks for the pull request, @kaustavb12!


When this pull request is ready, tag your edX technical lead.


<!-- comment:welcome-blended -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

kaustavb12 committed:
feat: merge ModuleSystem with DescriptorSystem

kaustavb12 committed:
feat: remove usage of LmsModuleSystem and PreviewModuleSystem

kaustavb12 committed:
fix: add namespace prefix to mako_service for redering in studio

kaustavb12 committed:
feat: remove CombinedSystem

kaustavb12 committed:
feat: removed StudioEditModuleRuntime

kaustavb12 committed:
chore: cleanup LmsModuleSystem and PreviewModuleSystem

kaustavb12 committed:
test: fix test cases and lint issues

kaustavb12 committed:
feat: add _runtime_services attribute to DescriptorSystem

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -313,19 +313,31 @@ class XModuleMixin(XModuleFields, XBlock):
+    @property
+    def xmodule_runtime(self):
+        """
+        Shim to mainatain backward compatibility.
```
```suggestion
        Shim to maintain backward compatibility.
```

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -313,19 +313,31 @@ class XModuleMixin(XModuleFields, XBlock):
+        return self._runtime
+
     @property
     def system(self):
```
Nit: could we add a deprecation warning to this one too?

kaustavb12 replied:
Added

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -313,19 +313,31 @@ class XModuleMixin(XModuleFields, XBlock):
+            'xmodule_runtime property is deprecated. Please use the runtime property instead.',
+            DeprecationWarning, stacklevel=3,
+        )
+        return self._runtime
```
Nit: shouldn't it use the `self.runtime` property, as described in the warning?

kaustavb12 replied:
Yes. Changed to correct property

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -608,8 +619,8 @@ def bind_for_student(self, xmodule_runtime, user_id, wrappers=None):
-            if getattr(xmodule_runtime, 'position', None):
-                self.position = xmodule_runtime.position   # update the position of the tab
+            if getattr(self._runtime, 'position', None):
+                self.position = self._runtime.position   # update the position of the tab
```
Nit: why aren't we using `self.runtime` directly?

kaustavb12 replied:
Changed to correct property

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -1321,7 +1070,7 @@ def anonymous_student_id(self):
             DeprecationWarning, stacklevel=3,
         )
-        user_service = self._services.get('user')
+        user_service = self._runtime_services.get('user') or self._services.get('user')
```
Why do we need `self._runtime_services`? Do we have a conflict with services from the XBlock Runtime?

kaustavb12 replied:
Yes, infact I became aware of this issue, due to [this test case](https://github.com/open-craft/edx-platform/blob/f543b6f9b44f8b60a55b39bf472cb0cdc212613d/lms/djangoapps/courseware/tests/test_block_render.py#L2725-L2755). It turns out, updating the services in `_services` also updates the services in other blocks and verticals.

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -1692,34 +1475,91 @@ def set(self, attr, val):
+    def get_block(self, usage_id, for_parent=None):
+        """See documentation for `xblock.runtime:Runtime.get_block`"""
+        block = self.load_item(usage_id, for_parent=for_parent)
+        if getattr(self, 'get_block_for_descriptor', None):
```
Please add a comment to explain why we need this.

kaustavb12 replied:
Added

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -1692,34 +1475,91 @@ def set(self, attr, val):
-        assert self.xmodule_instance is not None
-        return self.handler_url(self.xmodule_instance, 'xmodule_handler', '', '').rstrip('/?')
+        if getattr(self, 'applicable_aside_types_override', None):
+            return self.applicable_aside_types_override(block, applicable_aside_types=super().applicable_aside_types)
```
Please add a comment to explain why we need this override.

kaustavb12 replied:
Added

Agrendalath commented on xmodule/x_module.py:
```diff
@@ -1692,34 +1475,91 @@ def set(self, attr, val):
+        # that implements the correct handler url. So, for now, instead, we will reference a
+        # global function that the application can override.
+        if getattr(self, 'handler_url_override', None):
+            return self.handler_url_override(block, handler_name, suffix, query, thirdparty)
```
Please add a comment to explain why we need this override.

kaustavb12 replied:
Modified the existing comment

Agrendalath commented on cms/djangoapps/contentstore/views/block.py:
```diff
@@ -290,38 +290,23 @@ def can_write(self, course_key):
-    An extremely minimal ModuleSystem shim used for XBlock edits and studio_view.
-    (i.e. whenever we're not using PreviewModuleSystem.) This is required to make information
+    Function to set some required services used for XBlock edits and studio_view.
+    (i.e. whenever we're not loading preview module system.) This is required to make information
```
```suggestion
    (i.e. whenever we're not loading the `_preview_module_system`.) This is required to make information
```

Nit: it's hard to find any references to the "preview module system".

kaustavb12 replied:
Changed

Agrendalath commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -587,47 +587,49 @@ def inner_get_block(descriptor):
+    descriptor.runtime.get_block_for_descriptor = inner_get_block
+
+    # TODO: When we merge the descriptor and module systems, we can stop reaching into the mixologist (cpennington)
+    descriptor.runtime.mixins = descriptor.runtime.mixologist._mixins  # lint-amnesty, pylint: disable=protected-access
```
Is it still relevant?

kaustavb12 replied:
No, it isn't, since the runtime [already has those mixins](https://github.com/openedx/XBlock/blob/9cf2dfb1e33b2c73dc053b49e7a5cba3293db83b/xblock/runtime.py#L568)

Agrendalath commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -587,47 +587,49 @@ def inner_get_block(descriptor):
+    descriptor.runtime._runtime_services.update(services)  # lint-amnesty, pylint: disable=protected-access
+    descriptor.runtime.request_token = request_token
+    descriptor.runtime.wrap_asides_override = lms_wrappers_aside
+    descriptor.runtime.applicable_aside_types_override = lms_applicable_aside_types
```
Why are we using overrides instead of overriding methods directly (e.g., `descriptor.runtime.applicable_aside_types = lms_applicable_aside_types`)?

kaustavb12 replied:
This is purely to make debugging a bit easier in the future.

Debugging this code flow is pretty complicated as it is, and I felt it would become even more difficult if without any obvious indicator, the methods in the `DescriptorSystem` started behaving differently from what is originally defined. 

By adding these overrides, atleast the developer will know that there is a possible alternative execution path for this method.

Agrendalath commented on lms/djangoapps/lms_xblock/runtime.py:
```diff
@@ -72,6 +71,64 @@ def local_resource_url(block, uri):
+    runtime_class = 'LmsRuntime'
+    extra_data = {
+        'block-id': quote_slashes(str(block.scope_ids.usage_id)),
+        'course-id': quote_slashes(str(block.course_id)),
```
```suggestion
        'course-id': quote_slashes(str(block.scope_ids.usage_id.context_key)),
```

We have already deprecated `course_id`.

Agrendalath commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -437,7 +437,7 @@ def get_module_system_for_user(
 
     Returns:
-        (LmsModuleSystem, KvsFieldData):  (module system, student_data) bound to, primarily, the user and descriptor
+        KvsFieldData:  student_data bound to, primarily, the user and descriptor
```
Given this change, would it make sense to rename this to something like `prepare_runtime_for_user` or `setup_runtime_for_user`? It would also be great if we could rename `descriptor` to `block` here. Same for the `_preview_module_system` function.

Also, there are still some `ModuleSystem` references in the docstring and one comment.

kaustavb12 replied:
Changed:
`get_module_system_for_user` -> `prepare_runtime_for_user`
`_preview_module_system` -> `_prepare_runtime_for_preview`

Also, renamed `descriptor` to `block` in both the methods. However, I did not rename `descriptor` outside these two methods in [block_render.py](https://github.com/open-craft/edx-platform/blob/kaustav/unify_modulesystem_descriptorsystem/lms/djangoapps/courseware/block_render.py) and [preview.py](https://github.com/open-craft/edx-platform/blob/kaustav/unify_modulesystem_descriptorsystem/cms/djangoapps/contentstore/views/preview.py), such as in [get_block_for_descriptor()](https://github.com/open-craft/edx-platform/blob/f543b6f9b44f8b60a55b39bf472cb0cdc212613d/lms/djangoapps/courseware/block_render.py#L368).

Let me know if I should go ahead with changing those too.

Also removed references of `ModuleSystem` as much as possible from this and other files in the repo

Agrendalath commented on xmodule/services.py:
```diff
@@ -223,8 +222,7 @@ def rebind_noauth_module_to_user(self, block, real_user):
         # now bind the module to the new ModuleSystem instance and vice-versa
-        block.runtime = inner_system
-        inner_system.xmodule_instance = block
+        block.runtime.xmodule_instance = block
```
Haven't we removed `xmodule_instance` from `applicable_aside_types`? I don't see other references to this attribute in `edx-platform`.

kaustavb12 replied:
Thanks for pointing this out. I have removed this line, since this is redundant.

In `ModuleSystem`, `xmodule_instance` was used [here](https://github.com/openedx/edx-platform/blob/197fef9c2b31de27a5a462bcdd2d8e7f167a492c/xmodule/x_module.py#L1703-L1709) for `ajax_url` property.

While merging `ModuleSystem` and `DescriptorSystem`, I had removed this `ajax_url` property, since this was never referenced anywhere in the `edx-platform`. Instead, the `ajax_url` property was referenced from the [XModuleToXBlockMixin](https://github.com/openedx/edx-platform/blob/197fef9c2b31de27a5a462bcdd2d8e7f167a492c/xmodule/x_module.py#L764-L773) class, which is mixed-in to most XBlock classes.

However, this reference to `xmodule_instance` got me thinking if the `ajax_url` property was indeed referenced from somewhere outside `edx-platform`, and if I should add this property back to `DescriptorSystem`.

I'll see if I figure out a way to test the LTI codeflow, which seems to be the only place where the `xmodule_instance` is used.

kaustavb12 committed:
chore: address review comments

Agrendalath committed:
refactor: do not rely on `LmsModuleSystem` in Video XBlock

It was introduced in 159b707. We are removing this class.
This also adds a missing test case.

Agrendalath committed:
test: remove `only_xmodules`

This function was only used by tests. It was breaking tests from the `xmodule`
package on the devstack because the dist.key of its entry_points is `open-edx`.

Agrendalath reviewed (approved):
:+1:

* [x] I tested this: did the regression testing of XBlocks, instructor tasks, etc.
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation: n/a
* [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository: n/a

Agrendalath commented:
@ormsbee, this is ready for your review.

ormsbee commented on lms/djangoapps/courseware/block_render.py:
```diff
@@ -578,73 +577,73 @@ def inner_get_block(descriptor):
+    block.runtime.get_block_for_descriptor = inner_get_block
+
+    block.runtime.wrappers = block_wrappers
+    block.runtime._runtime_services.update(services)  # lint-amnesty, pylint: disable=protected-access
```
This pattern of updating the runtime services seems to happen pretty often. Does it make sense to make it a method on the runtime?

Agrendalath replied:
@ormsbee, excluding the tests (where we are mocking these services), this is used only in two places - `block_render.py` and `preview.py`. We are using `_runtime_services` only to avoid conflicts with `_services` from the XBlock Runtime. They should not be updated directly anywhere else, so keeping the direct private access and a manual `update` could make more sense here.

We also access `_runtime_services` and `_services` directly in the `ModuleSystemShim`, but these properties are deprecated and will be removed eventually.

Does it sound reasonable?

ormsbee commented on lms/djangoapps/courseware/tests/test_word_cloud.py:
```diff
@@ -1,14 +1,31 @@
+        block = self.load_item(usage_id, for_parent=for_parent)
+        return block
+
+    monkeypatch.setattr(DescriptorSystem, "get_block", get_block)
```
Optional: This pattern seems to be used repeatedly. Should it be a test helper somewhere in the `xmodule` package?

Agrendalath replied:
@ormsbee, sure, done. I added this fixture to `xmodule/tests/helpers.py`.

ormsbee reviewed (approved):
Great work folks!

Two optional items, but this looks good to me. <strike>I would like this to merge _after_ the Palm master branches are cut tomorrow. If there are subtle bugs or third party XBlock incompatibilities that result from these internal changes, I'd rather give more time to catch those issues before this hits a named release.</strike> **Edit (from email dated 2023-04-07): "Because of vacations on team events we'll be releasing to 2U stage on April 24th and, assuming no blockers, to production April 25th"** 

As always, please coordinate with @jristau1984 on precise merge timing. @jristau1984: You might want to broadcast this one pretty loudly internally, since it affects such a core part of courseware. You'll probably also want to check logs for the deprecation warnings added by this PR, in case edX is installing XBlocks that use these.

I'm personally interested to see if this reduces memory usage (I live in hope). 😄

ormsbee commented:
Sorry folks, I'm only now remembering there was an email about this with merge timing for April 24th, and prod on the 25th. I did not mean to override that.

e0d commented:
@jristau1984 this PR is was approved by @ormsbee, so we're all set for the release on 4/25.

Agrendalath committed:
chore: update DnDv2 XBlock

openedx-webhooks commented:
@kaustavb12 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/31472&entry.1484655318=https://github.com/kaustavb12&entry.752974735=2022-12-22+08:30&entry.1917517419=2023-04-24+14:03&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
